op-e2e/actions/proofs/l1_prague_fork_test.go

```diff
@@ -0,0 +1,164 @@
+				batcher.ActBufferAll(t)
+				batcher.ActL2ChannelClose(t)
+				batcher.ActSubmitSetCodeTx(t)
+			} else {
+				batcher.ActSubmitAll(t)
+			}
+			miner.ActL1IncludeTx(batcher.BatcherAddr)(t)
+			miner.ActL1EndBlock(t)
+		}
+
+		checkPragueStatusOnL1 := func(active bool) {
+			l1Head := miner.L1Chain().CurrentBlock()
+			if active {
+				// require.True(t, sd.L1Cfg.Config.IsPrague(l1Head.Number, l1Head.Time), "Prague should be active")
+				require.NotNil(t, l1Head.RequestsHash, "Prague header requests hash should be non-nil")
+			} else {
+				// require.False(t, sd.L1Cfg.Config.IsPrague(l1Head.Number, l1Head.Time), "Prague should not be active yet")
+				require.Nil(t, l1Head.RequestsHash, "Prague header requests hash should be nil")
+			}
+		}
+
+		syncVerifierAndCheck := func(t actionsHelpers.StatefulTesting) {
+			verifier.ActL1HeadSignal(t)
+			verifier.ActL2PipelineFull(t)
+			checkVerifierDerivedToL1Head(t)
+		}
+
+		// Check initially Prague is not activated
+		checkPragueStatusOnL1(false)
+
+		// Start op-nodes
+		sequencer.ActL2PipelineFull(t)
+		verifier.ActL2PipelineFull(t)
+
+		// Build L1 blocks, crossing the fork boundary
+		miner.ActEmptyBlock(t) // block 1
+		miner.ActEmptyBlock(t) // Prague activates here (block 2)
+
+		// Here's a block with a type 4 deposit transaction, sent to the OptimismPortal
+		miner.ActL1StartBlock(12)(t) // block 3
+		tx, err := actionsHelpers.PrepareSignedSetCodeTx(
+			*uint256.MustFromBig(env.Sd.L1Cfg.Config.ChainID),
+			env.Dp.Secrets.Alice,
+			env.Alice.L1.Signer(),
+			env.Alice.L1.PendingNonce(t), // nonce
+			env.Sd.DeploymentsL1.OptimismPortalProxy,
+			[]byte{})
+		require.NoError(t, err, "failed to prepare set code tx")
+		err = miner.EthClient().SendTransaction(t.Ctx(), tx)
+		require.NoError(t, err, "failed to send set code tx")
+		miner.ActL1IncludeTx(env.Alice.Address())(t)
+		miner.ActL1EndBlock(t)
+
+		// Check that Prague is active on L1
+		checkPragueStatusOnL1(true)
+
+		// Cache safe head before verifier sync
+		safeL2Before := verifier.SyncStatus().SafeL2
+
+		// Build L2 unsafe chain and batch it to L1 using either DynamicFee or
+		// EIP-7702 SetCode txs
+		// https://github.com/ethereum/EIPs/blob/master/EIPS/eip-7702.md
+		buildUnsafeL2AndSubmit(testCfg.Custom.useSetCodeTx)
+
+		// Check verifier derived from Prague L1 blocks
+		syncVerifierAndCheck(t)
+
+		// Check safe head did or did not change,
+		// depending on tx type used by batcher:
+		safeL2After := verifier.SyncStatus().SafeL2
+		if testCfg.Custom.useSetCodeTx {
+			require.Equal(t, safeL2Before, safeL2After, "safe head should not have changed (SetCode / type 4 batcher tx ignored)")
+			require.Equal(t, uint64(0), verifier.SyncStatus().SafeL2.L1Origin.Number, "l1 origin of l2 safe should not have changed (SetCode / type 4 batcher tx ignored)")
+		} else {
+			require.Greater(t, safeL2After.Number, safeL2Before.Number, "safe head should have progressed (DynamicFee / type 2 batcher tx derived from)")
```

Comment:
could we also add an explicit assertion about the new safe head number? e.g. that it's the same as the current unsafe?

Reply:
[9715058](https://github.com/ethereum-optimism/optimism/pull/14006/commits/97150584120a09b95c959821089fe58a9cfcf8a2)